dapr/proto/components/v1/state.proto

```diff
@@ -0,0 +1,212 @@
+import "dapr/proto/components/v1/common.proto";
+
+option go_package = "github.com/dapr/dapr/pkg/proto/components/v1;components";
+
+
+// QueriableStateStore service provides a gRPC interface for querier state store components.
+service QueriableStateStore {
+  // Query performs a query request on the statestore.
+  rpc Query(QueryRequest) returns (QueryResponse) {}
+}
+
+message Sorting {
+  // The key that should be used for sorting.
+  string key = 1;
+  enum Order {
+    ASC = 0;
+    DESC = 1;
+  }
+  // The order that should be used.
+  Order order = 2;
+}
+
+message Pagination {
+  // Maximum of results that should be returned.
+  int64 limit = 1;
+  // The pagination token.
+  string token = 2;
+}
+
+message Query {
+  // Filters that should be applied.
+  map<string, google.protobuf.Any> filter = 1;
+  // The sort order.
+  repeated Sorting sort = 2;
+  // The query pagination params.
+  Pagination pagination = 3;
+}
+
+// QueryRequest is for querying state store.
+message QueryRequest {
+  // The query to be performed.
+  Query query = 1;
+  // Request associated metadata.
+  map<string, string> metadata = 2;
+}
+
+// QueryItem is an object representing a single entry in query results.
+message QueryItem {
+  // The returned item Key.
+  string key = 1;
+  // The returned item Data.
+  bytes data = 2;
+  // The returned item ETag
+  dapr.proto.common.v1.Etag etag = 3;
+  // The returned error string.
+  string error = 4;
+  // The returned contenttype
+  string content_type = 5;
+}
+
+// QueryResponse is the query response.
+message QueryResponse {
+  // The query response items.
+  repeated QueryItem items = 1;
+  // The response token.
+  string token = 2;
+  // Response associated metadata.
+  map<string, string> metadata = 3;
+}
+
+// TransactionalStateStore service provides a gRPC interface for transactional state store components.
+service TransactionalStateStore {
+    // Multi executes multiples operation in a transactional environment.
+    rpc Multi(TransactionalStateRequest) returns (google.protobuf.Empty) {}
+}
+
+// TransactionalStateOperation describes operation type, key, and value for transactional operation.
+message TransactionalStateOperation {
+  // request is either delete or set.
+  oneof request {
+    DeleteRequest delete = 1;
+    SetRequest set = 2;
+  }
+}
+
+// TransactionalStateRequest describes a transactional operation against a state store that comprises multiple types of operations
+// The Request field is either a DeleteRequest or SetRequest.
+message TransactionalStateRequest {
+  // Operations that should be performed.
+  repeated TransactionalStateOperation operations = 1;
+  // Request associated metadata.
+  map<string, string> metadata = 2;
+}
+
+// StateStore service provides a gRPC interface for state store components.
+service StateStore {
+  // Initializes the state store component with the given metadata.
+  rpc Init(MetadataRequest) returns (google.protobuf.Empty) {}
```

Comment:
For future-proofing, I'd strongly recommend defining a type for each request or response, even if it's empty. For example, here you'd define messages `InitRequest` (`MetadataRequest` would be part of that) and `InitResponse` (which would be empty)

This way, we will be able to extend the proto in the future without breaking anything.

You can read more here: https://stackoverflow.com/a/50996412

Reply:
While I agree with you that this is cool, I think it ends up not having as much effect if you look at the whole context.

For this to change to occur, theoretically we would have to change the interface on code, which will lead us to change the entire chain of implementers/sdks, see that having an empty object in this case means not returning anything in the code, which will surely become a breaking changes anyway.

Another point is that from [what I noticed](https://github.com/dapr/dapr/blob/master/dapr/proto/runtime/v1/dapr.proto#L19), we already have this pattern in other APIs, is it interesting to deviate only here?

I'm not against it, but seems that in the other hand it could create empty objects without any strong reason,

unless we already have a strong intention to change an API soon, maybe it's a good one.

does it makes sense to you @ItalyPaleAle ?

Reply:
Yeah I see what you're saying. I guess it depends on how we think the API will evolve. Being able to add extra fields may allow us to make incremental evolutions without requiring an entire v2 version. Whether that's needed or not, perhaps should be a discussion.

Indeed we use that pattern everywhere in the code already. Still doesn't mean that we cannot use better patterns (if appropriate) in newer code.

Reply:
>Indeed we use that pattern everywhere in the code already. Still doesn't mean that we cannot use better patterns (if appropriate) in newer code.

Yes, I agree with you. The main question is, worth to pay API consistency vs non-predictable extensibility in this case?

Notice that changing the API is different from code. The API is what we expose as the interface for all SDKs, a change here will require a change in all SDKs, I generally care less about changes that cause inconsistency in the code if it is well scoped - new code for instance.

In the case of an API inconsistency - whether objects are empty or not - I think this should be taken as a decision at the API guidelines level and not punctually,

What do you think?

Reply:
Perhaps we should discuss this in a separate issue or during a call?

Reply:
I agree, where we should start it?

Reply:
As a followup discussion with @artursouza he convinced me that it is worth deviating here especially because the pluggable components proto is kinda new API that could change a lot soon. 

And even dough at the end of the day changing proto means changing the entire sdk chain, having empty messages will allow to break semantically but no structurally.

So, let's say we have a new field on `SetResponse` and this will require a new SDK version and a new component interface on code. For changes backwards compatible for instance, adding a new field on SetResponse, will also be semantically backwards compatible with components made on the previous version because the type didn't change. As opposite if it as a protobuf.empty, which will enforce components to update even if they don't want to use the new SetResponse field.

Thanks for suggesting this @ItalyPaleAle